kernel/src/log_segment.rs

```diff
@@ -311,31 +311,71 @@ fn list_log_files_with_version(
     let mut checkpoint_parts = vec![];
-    let mut max_checkpoint_version = start_version;
 
-    for parsed_path in list_log_files(fs_client, log_root, start_version, end_version)? {
-        let parsed_path = parsed_path?;
-        if parsed_path.is_commit() {
-            commit_files.push(parsed_path);
-        } else if parsed_path.is_checkpoint() {
-            let path_version = parsed_path.version;
-            match max_checkpoint_version {
-                None => {
-                    checkpoint_parts.push(parsed_path);
-                    max_checkpoint_version = Some(path_version);
+    let log_files = list_log_files(fs_client, log_root, start_version, end_version)?;
+
+    process_results(log_files, |iter| {
+        let log_files = iter.chunk_by(move |x| x.version);
+        for (version, files) in &log_files {
+            let mut new_checkpoint_parts = vec![];
+            for file in files {
+                if file.is_commit() {
+                    commit_files.push(file);
+                } else if file.is_checkpoint() {
+                    new_checkpoint_parts.push(file);
                 }
-                Some(checkpoint_version) => match path_version.cmp(&checkpoint_version) {
-                    Ordering::Greater => {
-                        max_checkpoint_version = Some(path_version);
-                        checkpoint_parts.clear();
-                        checkpoint_parts.push(parsed_path);
+            }
+
+            // Group checkpoint parts by the number of parts they have
+            let mut checkpoints = HashMap::new();
+            for part_file in new_checkpoint_parts {
+                use LogPathFileType::*;
+                match &part_file.file_type {
+                    SinglePartCheckpoint
+                    | UuidCheckpoint(_)
+                    | MultiPartCheckpoint {
+                        part_num: 1,
+                        num_parts: 1,
+                    } => {
+                        // All single-file checkpoints are equivalent, just keep one
+                        checkpoints.insert(1, vec![part_file]);
                     }
-                    Ordering::Equal => checkpoint_parts.push(parsed_path),
-                    Ordering::Less => {}
-                },
+                    MultiPartCheckpoint {
+                        part_num: 1,
+                        num_parts,
+                    } => {
+                        // Start a new multi-part checkpoint with at least 2 parts
+                        checkpoints.insert(*num_parts, vec![part_file]);
+                    }
+                    MultiPartCheckpoint {
+                        part_num,
+                        num_parts,
+                    } => {
+                        // Continue a new multi-part checkpoint with at least 2 parts
+                        if let Some(part_files) = checkpoints.get_mut(num_parts) {
+                            if *part_num == 1 + part_files.len() as u32 {
```

Comment:
Widening conversions are safer:
```suggestion
                            if *part_num as usize == 1 + part_files.len() {
```
(again below)

Reply:
nice catch!